Add `email` notifications channel

davelopez opened the pull request:
Closes #17882

This PR adds a new `email` channel to the notifications system.

## Backend changes
- Includes 2 database migrations to add 2 new columns to the `notification` table.
  - `dispatched` column: will determine if this notification has been "dispatched" through the additional communication channels. Any existing notification will set this value to "True" so users won't get emails retroactively.
  - `galaxy_url` column: allows to build URLs from the notification channel plugins that don't have access to the web request context.
- Move the creation of notifications (database objects) to a celery task if enabled. This can offload resolving the recipient users when we include complex combinations of groups and/or roles as recipients. We can also control if we want to explicitly run this synchronously when we don't use groups or roles.
- Adds a new periodic celery task (every 10 minutes by default) that will process non `dispatched` notifications and send them via all available channels. In-app notifications don't depend on this task so they will pop up in Galaxy at their due time. Other external channels, like email, will probably carry some delay depending on the `dispatch_notifications_interval` configuration option.
- Adds personalizable email templates for the currently existing types of notifications.

## Frontend changes
- The notification preferences will display a new `Email` channel if the server "supports" it. The channels supported by the server are passed to the client via a new header named `supported-channels` when requesting `api/notifications/preferences`. 
- Navigating to `/user/notifications?preferences=true` will display the preferences panel open. This makes it easier to link to the preferences from the email templates.

## Sample Emails

### Shared Item
![Screenshot from 2024-04-11 13-46-29](https://github.com/galaxyproject/galaxy/assets/46503462/b907b1c2-99b1-4b17-afb2-4e6c49451e7d)


### Direct Message
![Screenshot from 2024-04-11 13-49-10](https://github.com/galaxyproject/galaxy/assets/46503462/4b36b2dd-6cad-4668-b680-4c1ec2e32828)

## See it in action
![NotificationEmails](https://github.com/galaxyproject/galaxy/assets/46503462/f982f132-5ce0-43b7-bc67-89a91c7bb7d1)


## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Make sure `enable_celery_tasks` is set to true in your Galaxy config.
  - Set up the `smtp_server` in your config. I used https://github.com/rnwood/smtp4dev/
  - Play around with the email templates from `lib/galaxy/config/templates/mail/notifications/` by copying them and setting the `templates_dir` accordingly.
  - Play around disabling the email channel for certain notifications in your preferences.
  - Play around scheduling notifications as an admin.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Allow to send notifications from Celery

Adds a new task to Celery called `send_notification_to_recipients_async`

davelopez committed:
Use NotificationService instead of manager in other services

Allows to use the asynchronous sending of notifications from other services.

davelopez committed:
Update NotificationCreateRequest definition

Make sure the API endpoint supports the asynchronous mode too.

davelopez committed:
Add email templates for message notifications

davelopez committed:
Add email templates for shared item notifications

davelopez committed:
Add EmailNotificationChannelPlugin and basic email sending

davelopez committed:
Add db revision for notifications dispatched column

davelopez committed:
Add dispatched column to notification model

davelopez committed:
Adapt notification model mapped types

Unfortunately I thought the default for columns was "non-nullable" so these columns are nullable in the database, but they will never be in the app.

This change matches the type expected in the app with the real nullable state in the database.

davelopez committed:
Ignore some inaccurate types from model mappings

davelopez committed:
Refactor notifications manager to send via other channels

This will still create the notifications and associations in the database on request, but will defer sending emails (or any other channel) to a different function called `dispatch_pending_notifications_via_channels`.

davelopez committed:
Add celery task to dispatch pending notifications

davelopez committed:
Add dispatch_notifications_interval config option

davelopez committed:
Add dispatch_pending_notifications to periodic tasks

davelopez committed:
Schedule periodic notifications tasks only when enabled

davelopez committed:
Run notification integration tests without celery

davelopez committed:
Revert utcnow refactoring

Is not yet deprecated in Python 3.7

davelopez committed:
Update client schema

davelopez committed:
Handle channel settings depending on server config

External notification channels (like email) are processed in Celery tasks. When Celery is not enabled in the server, we cannot show those channels to the user and we have to indicate in the API that those will be ignored.

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -2939,13 +2939,18 @@ class Notification(Base, Dictifiable, RepresentById):
+        String(16), index=True, nullable=True
     )  # Defines the 'importance' of the notification ('info', 'warning', 'urgent', etc.). Used for filtering, highlight rendering, etc
+    dispatched: Mapped[Boolean] = mapped_column(
+        Boolean, index=True, nullable=False, default=False
```
```suggestion
        Boolean, index=True, default=False
```
No need: derived from the type hint

davelopez committed:
Remove redundant nullable in dispatched column

Co-authored-by: John Davis <jdavcs@gmail.com>

davelopez committed:
Fix generic types in notifications schema

Workaround issues with generics auto-generated names when building the client schema.

davelopez committed:
Refactor `GenericModel` for reusability

davelopez committed:
Allow to `force_sync` for creating internal notifications

This is useful when we know in advance the list of recipient users and that no group or roles need to be resolved. Like when we share an item with a list of users.

davelopez committed:
Adapt notification integration tests to work w/o tasks

davelopez committed:
Add db migration for galaxy_url column in notifications table

davelopez committed:
Add galaxy_url mapped column to notification model

davelopez committed:
Update notification creation and sharing logic

This commit updates the notification creation and sharing logic to include the `galaxy_url` field in the `NotificationCreateRequest` and `SharedItemNotificationFactory` classes. This field is used to generate links in the notification content and is passed as an argument when sending notifications to recipients.

davelopez committed:
Update notification email templates

davelopez committed:
Open notification preferences if specified in route query

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -2939,13 +2939,21 @@ class Notification(Base, Dictifiable, RepresentById):
+    variant: Mapped[str] = mapped_column(
+        String(16), index=True, nullable=True
     )  # Defines the 'importance' of the notification ('info', 'warning', 'urgent', etc.). Used for filtering, highlight rendering, etc
+    dispatched: Mapped[Boolean] = mapped_column(
```
```suggestion
    dispatched: Mapped[bool] = mapped_column(
```
we need the Python data type here, not the SQLAlchemy type (SA2.0 maps Python's `bool` to `Boolean` automatically)

davelopez replied:
good catch! thanks!

davelopez committed:
Use proper type for Mapped

Co-authored-by: John Davis <jdavcs@gmail.com>

jdavcs commented on lib/galaxy/config/templates/mail/notifications/message-email.html:
```diff
@@ -0,0 +1,84 @@
+Use this template to customize the HTML email your users will receive
```
```suggestion
Use this template to customize the HTML-formatted email your users will receive
```
minor! up to you.

jdavcs commented on lib/galaxy/config/templates/mail/notifications/message-email.html:
```diff
@@ -0,0 +1,84 @@
+when a new notification of category "message" is sent to them.
+Copy the file to {{ templates_dir }}/mail/notifications/message-email.html and modify as required.
+
+If you are adding URLs into this, remember that only absolute URLS (with
```
```suggestion
If you are adding URLs, remember that only absolute URLs (with
```

jdavcs commented on lib/galaxy/config/templates/mail/notifications/message-email.txt:
```diff
@@ -0,0 +1,49 @@
+when a new notification of category "message" is sent to them.
+Copy the file to {{ templates_dir }}/mail/notifications/message-email.txt and modify as required.
+
+If you are adding URLs into this, remember that only absolute URLS (with
```
```suggestion
If you are adding URLs, remember that only absolute URLs (with
```

jdavcs commented on lib/galaxy/config/templates/mail/notifications/new_shared_item-email.html:
```diff
@@ -0,0 +1,80 @@
+when a new notification of category "new_shared_item" is sent to them.
+Copy the file to {{ templates_dir }}/mail/notifications/new_shared_item-email.html and modify as required.
+
+If you are adding URLs into this, remember that only absolute URLS (with
```
```suggestion
If you are adding URLs, remember that only absolute URLs (with
```

jdavcs commented on lib/galaxy/config/templates/mail/notifications/new_shared_item-email.txt:
```diff
@@ -0,0 +1,50 @@
+when a new notification of category "new_shared_item" is sent to them.
+Copy the file to {{ templates_dir }}/mail/notifications/new_shared_item-email.txt and modify as required.
+
+If you are adding URLs into this, remember that only absolute URLS (with
```
```suggestion
If you are adding URLs, remember that only absolute URLs (with
```
one more! :)

jdavcs commented on lib/galaxy/config/templates/mail/notifications/message-email.html:
```diff
@@ -0,0 +1,84 @@
+a domain name) make sense in email! They can be served from any stable
+location, including your Galaxy server or GitHub.
+
+The following variables are available for inserting into the html with Jinja2
```
```suggestion
The following variables are available for inserting into the HTML with Jinja2
```

jdavcs commented on lib/galaxy/config/templates/mail/notifications/new_shared_item-email.html:
```diff
@@ -0,0 +1,80 @@
+a domain name) make sense in email! They can be served from any stable
+location, including your Galaxy server or GitHub.
+
+The following variables are available for inserting into the html with Jinja2
```
```suggestion
The following variables are available for inserting into the HTML with Jinja2
```

jdavcs commented on lib/galaxy/managers/notification.py:
```diff
@@ -110,31 +151,118 @@ def ensure_notifications_enabled(self):
+
+        # Mark all pending notifications as dispatched
+        for notification in pending_notifications:
+            notification.dispatched = True
```
but what if an error happens in the next loop and a notification is not dispatched via `_dispatch_notification_to_users`? Then its state in the database would be incorrect, right?

jdavcs replied:
..Or is the notification dispatched/not dispatched state decoupled from successful dispatch to users?

davelopez replied:
I thought a bit about that, and I chose not to enforce a successful dispatch to all users (maybe using retries) because it feels "less annoying" from the user's perspective to receive no email than to receive it potentially multiple times...

Thinking about it again now (thanks to your comment :laughing:), maybe it would be better to move the dispatched column to the UserNotificationAssociation and have more fine control over which users the notification was "dispatched" successfully and for which it failed :thinking: Then we could think how to handle the second try for those that failed... What do you think?

jdavcs replied:
It makes sense, I think. The downside is we no longer have a quick way to determine if a notification has been dispatched (checking status of each notification based on all existing user associations is a no-go: performance + no easy way to handle new users who would qualify for a given notification, but where created after the notification was sent - it gets hairy..). We could have a status field on the Notification, set only when all users have been processed. However, if "has been processed" is defined as "task completed without error" then it's safe, I think, since it's dependent only on our code not raising an error. But if it's any more detailed (has been sent or has been received, etc.), we may end up with lots of "pending" notifications because of a problem dispatching to one user out of thousands.

I think the main question is how do we want to handle a failed notification dispatch. Sorry, maybe it's simpler than I think and all these are non-issues..

davelopez replied:
Then, probably the current approach is safe enough. This is what happens now:
### Phase 1: Sending the notification
1. An admin sends a notification to X users, Y groups, and/or Z roles
2. The `Notification` is created with `dispatched=false`
3. A `UserNotificationAssociation` is created for each user after resolving X, Y, and Z into a flat set of users (only those who didn't opt out from that notification)
4. At this point, the users can see their notifications in Galaxy (unless the notification was scheduled)

### Phase 2: Dispatching notifications to other communication channels
5. A periodic task (every 10 minutes by default) tries to dispatch all "pending" notifications (dispatched=False)
6. The task grabs the currently pending notifications and sets them as dispatched
7. For each notification
8. For each associated user
9. Send a notification through every channel (currently only send an email)

The idea behind `step 6`, marking them as dispatched before processing, is to avoid a second task grabbing some of those notifications and duplicating emails (etc.) if, for some reason, sending the emails takes longer than the periodic check.

If sending an email fails, it will not affect other users, but there will be no retry and the email will be lost. I think that is preferable over getting multiple potential emails.

jdavcs replied:
Yes, I think we can leave as is.  I think, it makes sense especially given items 1.4 and 2.9: if a notification is displayed to the user (1.4) it feels sufficient to refer to it as dispatched; and, given that we may have multiple channels, tracking successful dispatches for each channel is a whole lot of additional logic I hope we'll never have to implement in the context of Galaxy.. 

Thank you for the detailed description!

davelopez committed:
Apply suggestions from code review

Co-authored-by: John Davis <jdavcs@gmail.com>

jdavcs reviewed (approved):
Looks great! Thank you!
